### PR TITLE
Precompute ancestor hierarchy to speed up term indexing

### LINKED ISSUE #278: Term indexing extremely slow for large ontologies due to per-class SPARQL ancestor traversal
## Summary

Term indexing performance degrades severely for large ontologies. The root cause is that `index_doc` calls `retrieve_hierarchy_ids(:ancestors)` individually for every class, issuing iterative SPARQL queries level-by-level to walk up the hierarchy. For an ontology with 100K+ classes at average depth ~8, this produces hundreds of thousands of SPARQL round-trips just to populate the `parents` field in Solr.

## Current behavior

During `OntologySubmissionIndexer#index`, for each batch of 2,500 classes:
1. `Class.indexBatch(page_classes)` is called
2. This calls `indexable_object` → `index_doc` on every class
3. Each `index_doc` calls `retrieve_hierarchy_ids(:ancestors)` which iterates level-by-level up to 40 levels, issuing a SPARQL query at each level
4. For a class at depth D, that's D SPARQL round-trips

Total cost: O(N × avg_depth) SPARQL queries, where N is the number of classes.

## Proposed fix

Replace the per-class SPARQL ancestor traversal with a bulk precomputation:

1. Before the indexing loop, fetch **all** parent-child edges in a single paginated SPARQL query
2. Build the transitive closure in memory using memoized BFS — each edge visited at most once, total work O(V + E)
3. Store the precomputed ancestor map as a class-level cache on `LinkedData::Models::Class` for the duration of bulk indexing
4. `index_doc` reads ancestors from the cache instead of issuing SPARQL queries
5. Clear the cache after indexing completes (ensure block)

This replaces ~800K SPARQL round-trips with 1 paginated SPARQL query + in-memory computation.

## Files involved

- `lib/ontologies_linked_data/services/submission_process/operations/submission_indexer.rb` — indexing orchestration
- `lib/ontologies_linked_data/models/class.rb` — `index_doc`, `retrieve_hierarchy_ids`

## Additional context

There are secondary performance bottlenecks in the indexer (SPARQL page fetch inside sync block, `bring_remaining` in CSV writer under lock) that could be addressed in follow-up work.

### PR DESCRIPTION
**Pull request by alexskr**
## Summary

Closes #278

- Replace per-class SPARQL ancestor traversal in `index_doc` with a single bulk query + in-memory transitive closure
- Before the indexing loop, fetch all parent-child edges in one paginated SPARQL query and compute ancestors via memoized BFS — O(V + E) total
- Store the precomputed ancestor map as a class-level cache on `LinkedData::Models::Class`, cleared in an `ensure` block after indexing
- For a 100K-class ontology, this eliminates ~800K SPARQL round-trips
- Observed **2-5x overall indexing performance improvement** on large ontologies

## Changes

- **`submission_indexer.rb`** — `compute_ancestors_map`, `fetch_all_parent_edges`, `compute_ancestors_for` precompute the full ancestor map; `validate_class_ancestors` provides temporary per-class comparison against the old traversal, gated behind `OP_VALIDATE_ANCESTORS` env var, with per-class timing logs
- **`class.rb`** — `ancestors_cache` class accessor; `index_doc` reads from cache instead of calling `retrieve_hierarchy_ids`
- **`test_ancestors_precompute.rb`** — 9 unit tests covering linear chain, diamond inheritance, multiple roots, cycles, complex DAG, memoization, and edge cases

## Validation

Run with `OP_VALIDATE_ANCESTORS=1` to enable per-class comparison of precomputed ancestors against the old SPARQL traversal. Every class is checked and logs include timing for old vs new approach plus match/mismatch status. Once validated against production data, the validation code can be removed.

## Test plan

- [x] Unit tests pass: `bundle exec rake test:docker:fs TEST='test/models/test_ancestors_precompute.rb'` (9/9)
- [ ] Full indexing pipeline: `bundle exec rake test:docker:fs TEST='test/models/test_search.rb'`
- [ ] Validation run on a large ontology with `OP_VALIDATE_ANCESTORS=1` — confirm zero mismatches

### CODE REVIEW

File: lib/ontologies_linked_data/models/class.rb [hunk 1]
```diff
@@ -254,15 +262,17 @@ def index_doc(to_set=nil)
...
+            # TODO: do we ever need per-class ancestor lookup outside of bulk indexing?
+            # If so, uncomment the fallback below.
+            # if self.class.ancestors_cache
+            #   path_ids = (self.class.ancestors_cache[class_id] || Set.new).dup
+            # else
+            #   path_ids = retrieve_hierarchy_ids(:ancestors)
             # end
-            # path_ids.delete(class_id)
-            path_ids = retrieve_hierarchy_ids(:ancestors)
+            path_ids = (self.class.ancestors_cache[class_id] || Set.new).dup
```
Comment (mdorf) [inline lib/ontologies_linked_data/models/class.rb:272]: ### **Hard dependency on cache — no fallback** (High)

`class.rb:270` now unconditionally reads from `ancestors_cache`:

```ruby
path_ids = (self.class.ancestors_cache[class_id] || Set.new).dup
```

If `index_doc` is ever called outside of bulk indexing (e.g., individual class re-indexing, tests, or other code paths), `ancestors_cache` will be `nil`, and calling `[class_id]` on `nil` will raise `NoMethodError`. The commented-out fallback to `retrieve_hierarchy_ids` should be uncommented:

```ruby
if self.class.ancestors_cache
  path_ids = (self.class.ancestors_cache[class_id] || Set.new).dup
else
  path_ids = retrieve_hierarchy_ids(:ancestors)
end
```

File: lib/ontologies_linked_data/models/class.rb [hunk 2]
```diff
@@ -15,6 +15,14 @@ class Class < LinkedData::Models::Base
       include LinkedData::Concerns::Concept::InScheme
       include LinkedData::Concerns::Concept::InCollection
 
+      class << self
+        # When set, index_doc uses this precomputed map instead of per-class
+        # SPARQL ancestor traversal. Keyed by class URI string, values are
+        # Sets of ancestor URI strings. Set by OntologySubmissionIndexer
+        # during bulk indexing and cleared after completion.
+        attr_accessor :ancestors_cache
```
Comment (mdorf) [inline lib/ontologies_linked_data/models/class.rb:23]: ### **Thread safety of class-level mutable state** (High)

`ancestors_cache` is a class-level `attr_accessor` on `LinkedData::Models::Class`. If two ontologies are indexed concurrently (separate threads/workers), one worker could overwrite or `nil` out the cache while another is still reading it. Consider:
- Making the cache per-submission (pass it through the indexing context rather than a global class variable)
- Or at minimum, documenting/asserting that concurrent indexing is not supported

File: lib/ontologies_linked_data/services/submission_process/operations/submission_indexer.rb [hunk 3]
```diff
@@ -194,6 +201,104 @@ def index(logger, commit = true, optimize = true)
...
+              queue.push(grandparent) unless visited.include?(grandparent)
+            end
+          end
+        end
+
+        ancestors_map[cls] = visited
+      end
+
+      # TODO: Remove once precomputed ancestors are validated against production data
+      def validate_class_ancestors(cls, logger)
```
Comment (mdorf) [inline lib/ontologies_linked_data/services/submission_process/operations/submission_indexer.rb:277]: ### **Instance variables leaked from `validate_class_ancestors`** (Medium)

In `validate_class_ancestors`, `@old_ancestors_result` and `@new_ancestors_result` are instance variables on the indexer but are only used within `Benchmark.realtime` blocks. These should be local variables instead. The current approach leaks state across calls:

```ruby
old_time = Benchmark.realtime do
  old_ancestors = cls.retrieve_hierarchy_ids(:ancestors)
  old_ancestors.select! { |x| !x["owl#Thing"] }
end
```

**Review by mdorf — COMMENTED** [review]
This is a SOLID performance optimization, and we should definitely implement it in a short working order. However, it introduces significant changes to a core model and should NOT be bundled together with the upcoming major release. I propose that we revisit this feature SHORTLY after the main release.

File: lib/ontologies_linked_data/models/class.rb (hunk 1 above)
Comment (alexskr) [reply lib/ontologies_linked_data/models/class.rb:272]: I was thinking about it but is there a case where index_doc is called outside of bulk indexing?

File: lib/ontologies_linked_data/models/class.rb (hunk 2 above)
Comment (alexskr) [reply lib/ontologies_linked_data/models/class.rb:23]: > ### Thread safety of class-level mutable state (High)

The `ancestors_cache` is computed once before worker threads are spawned. During indexing, threads only read from it (hash lookups + `.dup` on the Sets) — no writes occur until all threads complete and the cache is cleared.

The class-level state concern would apply if different ontologies were indexed concurrently in the same process. Currently ncbo_cron processes ontologies sequentially, and if we ever move to concurrent indexing, we would likely use separate worker processes subscribing to a queue — each with their own isolated class-level state. In the unlikely event we needed concurrency within a shared process, we could scope the cache via `Thread.current` (same pattern already used for `RequestStore.store[:requested_lang]` in this file).

File: lib/ontologies_linked_data/services/submission_process/operations/submission_indexer.rb (hunk 3 above)
Comment (alexskr) [reply lib/ontologies_linked_data/services/submission_process/operations/submission_indexer.rb:277]: > ### Instance variables leaked from `validate_class_ancestors` (Medium)

Good catch. Fixed in 5ffb7b65 — replaced with local variables and renamed to `sparql_ancestors`/`cached_ancestors` for clarity. Note this is a temporary validation method gated behind `OP_VALIDATE_ANCESTORS` that will be removed once we confirm correctness against production data.
